Move Native Library APIs to NativeLibrary class

swaroop-sridhar opened the pull request:
Move APIs that manipulate native libraries from
System.Runtime.InteropServices.Marshal class to the new
System.Runtime.InteropServices.NativeLibrary class.

In API review: #32015,
The LoadLibrary APIs were originally Approved to live in
System.Runtime.InteropServices.Marshal class.

However, recently the decision was changed such that the APIs are
in a new NativeLibrary class.

jkotas commented on src/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs:
```diff
@@ -0,0 +1,178 @@
+        /// <exception cref="System.ArgumentNullException">If libraryPath is null</exception>
+        /// <exception cref="System.DllNotFoundException ">If the library can't be found.</exception>
+        /// <exception cref="System.BadImageFormatException">If the library is not valid.</exception>
+        public static IntPtr LoadLibrary(string libraryPath)
```
Is `Library` still needed in the method names now that the `Library` is in the type name? "Library" in the 
method names looks superfluous:

```csharp
var handle = NativeLibrary.LoadLibrary("foo")
NativeLibrary.GetLibraryExport(handle, "m");
```

jkotas commented on src/vm/nativelibrarynative.cpp:
```diff
@@ -0,0 +1,78 @@
+//
+
+//
+// FCall's for the PInvoke classlibs
```
Nit: This comment looks useless.

AaronRobinsonMSFT commented on src/vm/nativelibrarynative.h:
```diff
@@ -0,0 +1,29 @@
+//
+
+//
+// QCall's for the NativeLibrary class
```
Superfluous.

swaroop-sridhar replied:
I agre; but I followed the existing convention in other native.h files https://github.com/dotnet/coreclr/blob/master/src/vm/marshalnative.h#L9
https://github.com/dotnet/coreclr/blob/master/src/vm/assemblynative.hpp#L9

AaronRobinsonMSFT commented on src/System.Private.CoreLib/src/System/Runtime/InteropServices/NativeLibrary.cs:
```diff
@@ -0,0 +1,178 @@
+        internal static extern IntPtr LoadLibraryFromPath(string libraryName, bool throwOnError);
+        [DllImport(JitHelpers.QCall, CharSet = CharSet.Unicode)]
+        internal static extern IntPtr LoadLibraryByName(string libraryName, RuntimeAssembly callingAssembly, 
+                                                        bool hasDllImportSearchPathFlag, uint dllImportSearchPathFlag, 
```
In this case, I think you want to specify the marshalling of the `bool`s. In this case, you are using C# `bool` which is 1 byte, but in the native signature below the Win32 `BOOL` value is being used which is 4-bytes. I can't recall the default off the top of my head, but I would suggest being explicit here - i.e. `[MarshalAs(UnmanagedType.Bool)]`

jkotas replied:
FWIW, there are many QCalls with bool arguments - none of them have this extra ceremony.

swaroop-sridhar replied:
OK https://github.com/dotnet/coreclr/pull/21821/commits/f1933b7b04ee4e09c7e335761f0b552bdce26ebd Revert the annotations, and squash to one commit.

AaronRobinsonMSFT replied:
Sounds good.

swaroop-sridhar commented:
Thanks for the feedback @AaronRobinsonMSFT  and @jkotas 
I've addressed these comments in https://github.com/dotnet/coreclr/pull/21821/commits/b20b30e14e27f61aaf5692ffb71a42e00f664029

jkotas commented:
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test please
@dotnet-bot test Ubuntu arm Cross Checked Innerloop Build and Test please
